Feat: Toxin damage applies on organs on death

fake123acc opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Dying with non-zero toxin damage results in harm to the liver, in case of liver failure, it also damages the heart and other inner organs. The more toxin damage the more organs suffer
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Increasing significance of the base damage type which didn't prevent from using a defibrillator and usually was negated on heart restart.
https://discord.com/channels/617003227182792704/755125334097133628/975109761089552484
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif of your feature if you want -->

## Changelog
:cl: nopeucant
add: Added damage on internal organs in case you die with toxin damage on character. The more toxins the more organs suffer (heart necrosis on 120 dmg, total internal organs failure on 270 dmg)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

fake123acc committed:
Feat: Toxin damage applies on organs on death

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Dying with non-zero toxin damage results harm to liver, in case of liver failure it also damages heart and other inner organs. The more toxin damage the more organs suffer
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Increasing significance of the base damage type which didn't prevent from using defibrillator and usually was negated on heart restart.
https://discord.com/channels/617003227182792704/755125334097133628/975109761089552484
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif of your feature if you want -->

## Changelog
:cl: nopeucant
add: Added damage on internal organs in case you die with toxin damage on character. The more toxins the more organs suffer (heart necrosis on 120 dmg, total internal organs failure on 270 dmg)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

fake123acc commented:
Немного иллюстраций "до-после". 
 

Всем сканам за исключением тех, что соответствуют 200+ мгновенного урона токсинами, предшествовало не менее полутора минут (как правило больше, в случае с радиацией - от 5 до 10) рвоты, шатания и обмороков с рыданиями (а в случае с зарином - ещё и криков).

<details>
<summary>100 токсинов</summary>

[![100 токсинов][1]][1]

[1]: https://user-images.githubusercontent.com/105564991/168454415-c7ac0ac1-0b20-4ec6-b5ac-7e1f7e47b233.PNG
</details>
<details>
<summary>150 токсинов</summary>

[![150 токсинов][2]][2]

[2]: https://user-images.githubusercontent.com/105564991/168454419-3aad8059-fb5f-433f-9709-87deae762af8.PNG
</details>
<details>
<summary>200 токсинов</summary>

[![200 токсинов][3]][3]

[3]: https://user-images.githubusercontent.com/105564991/168454421-235d4d30-7214-489a-98ce-c77c265171f5.PNG
</details>
<details>
<summary>250 токсинов</summary>

[![250 токсинов][4]][4]

[4]: https://user-images.githubusercontent.com/105564991/168454409-d112fda2-1f47-4479-84f8-032a59fedfd2.PNG
</details>
<details>
<summary>300 токсинов</summary>

[![300 токсинов][5]][5]

[5]: https://user-images.githubusercontent.com/105564991/168454410-c2a809b7-9b55-4d1b-97f0-6b5cb1ddf7ca.PNG
</details>

<details>
<summary>120 радиации</summary>

[![120 радиации][6]][6]

[6]: https://user-images.githubusercontent.com/105564991/168454416-03fccab4-d6fc-4fdb-ba9f-7cb7e78f3afd.PNG
</details>
<details>
<summary>150 радиации</summary>

[![150 радиации][7]][7]

[7]: https://user-images.githubusercontent.com/105564991/168454417-a1c65cda-64cf-46c2-9ef0-99646ff315c5.PNG
</details>
<details>
<summary>200 радиации</summary>

[![200 радиации][8]][8]

[8]: https://user-images.githubusercontent.com/105564991/168454420-99cee2de-2231-4cc3-85a1-a765db4136e9.PNG
</details>

<details>
<summary>15 зарина</summary>

[![15 зарина][9]][9]

[9]: https://user-images.githubusercontent.com/105564991/168454414-c77d896a-918d-4ecb-894f-530352476f97.PNG
</details>
<details>
<summary>7 зарина, 7 цианида, 14 панкурония</summary>

[![17 зарина, 7 цианида, 14 панкурония][10]][10]

[10]: https://user-images.githubusercontent.com/105564991/168454413-a9e9c930-8896-4b45-87d3-327168431331.PNG
</details>

<details>
<summary>Был избит ядовитым пауком до крита, дальше лежал пока не умер</summary>

[![Был избит ядовитым пауком до крита, дальше лежал пока не умер][11]][11]

[11]: https://user-images.githubusercontent.com/105564991/168454411-27146246-91c7-4405-9993-6901780a9653.PNG
</details>
<details>
<summary>Дрался с ядовитым пауком телескопичкой, без брони. После его убийства две минуты стоял.</summary>

[![Дрался с ядовитым пауком телескопичкой, без брони. После его убийства две минуты стоял.][12]][12]

[12]: https://user-images.githubusercontent.com/105564991/168454412-19f54ea4-f857-4709-b5f9-5174e81b8be2.PNG
</details>

Vallat commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,36 @@
+			to_be_picked_from.Remove(organ_picked)
+
+
+/mob/living/carbon/proc/__apply_damage(var/dmg, var/obj/item/organ/targeted_organ) // only ment to be used in function above
+	var/can_be_absorbed = targeted_organ.max_damage - targeted_organ.damage
+	var/damage_to_be_applied = min(can_be_absorbed, dmg)
+	targeted_organ.receive_damage(damage_to_be_applied)
+	return damage_to_be_applied // returns damage wich was actually applied
```
Если эта функция нужна только чтобы "сократить" количество повторяющегося кода для функции `process_toxin_damage_on_death`, то просто немного видоизмени логику и вынеси её в макрос. Для этого вовсе не обязательно клепать функции

fake123acc replied:
Это мог бы быть вообще one-liner, и насрать, что повторяется. 
Но я подумал, что хочу читаемости.
Это плохо?

Vallat replied:
Читаемость для таких вещей как раз и создается через макросы, а не через функции.

Vallat commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,36 @@
+#define TOXIN_TO_INTERNAL_DAMAGE_MULTIPLIER 2
```
Лучше перенести это куда-нибудь в `__DEFINES/mobs.dm`, например

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,36 @@
+#define TOXIN_TO_INTERNAL_DAMAGE_MULTIPLIER 2
+
+// This function applies damage to internal organs in case the mob died having toxin damage
+// This feature increases amount of effort it takes to revive a victim of poisoning
```
Используй такой комментарий на проках как минимум. Эти комментарии будет подсвечивать, если установлен dm-langserver и кто-то наведет мышкой на прок, там где он вызывается.
Ну и вообще так принятно комментить. 
```suggestion
/**
* This function applies damage to internal organs in case the mob died having toxin damage
* This feature increases amount of effort it takes to revive a victim of poisoning
*/
```

fake123acc committed:
Moved define from mob/living/carbon/death.dm

fake123acc committed:
Moved define to anther file, changed apply_dmg fun

TOXIN_TO_INTERNAL_DAMAGE_MULTIPLIER was moved to _DEFINES
__apply_damage function was turned into APPLY_DAMAGE macro

fake123acc committed:
Applied Gaxxer note on comments

Co-authored-by: Gaxeer <44334376+Gaxeer@users.noreply.github.com>

fake123acc committed:
Fixed comments on macro

Changed em in order to keep it all in one style

Vallat reviewed (approved):
Вроде нормально. Надеюсь `while` не уйдет в бесконечный цикл :)

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+*/
+#define APPLY_DAMAGE(_damage, _part, _res) \
+do {\
+	var/can_be_absorbed = _part.max_damage - _part.damage;\
+	var/damage_to_be_applied = min(can_be_absorbed, _damage);\
```
```suggestion
	var/can_be_applied = _part.max_damage - _part.damage;\
	var/damage_to_be_applied = min(can_be_applied, _damage);\
```

Gaxeer replied:
Мне кажется, что по логике этого макроса, такое название подходит больше. Absorb - это в переводе `поглощение', и возникает ассоциация, что орган "впитывает" урон, а не просто высчитывается сколько он может получить до своей смерти

fake123acc replied:
Это именно то, что я имел в виду. Я считаю, сколько урона "впитывает" орган, принимая на себя. Если необходимо - приму предложение, но искренне, на 100% убеждён, что мой вариант лучше.

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+do {\
+	var/can_be_absorbed = _part.max_damage - _part.damage;\
+	var/damage_to_be_applied = min(can_be_absorbed, _damage);\
+	_part.receive_damage(damage_to_be_applied);\
```
Можем словить рантайм тут

fake123acc replied:
Знаки вопроса поставить?

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+*/
+#define APPLY_DAMAGE(_damage, _part, _res) \
+do {\
+	var/can_be_absorbed = _part.max_damage - _part.damage;\
```
и тут тоже может быть рантайм

fake123acc replied:
Точно может? У меня макрос задефайнен только на эти 30 строчек. Что в них может пойти не так, чтобы сломать макрос?

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+
+	// now the rest of the damage is going to be distributed among other organs
+	var/list/to_be_picked_from = target.internal_organs.Copy()
+	while (internal_damage > 0 && to_be_picked_from.len > 0)
```
```suggestion
	while (internal_damage > 0 && length(to_be_picked_from) > 0)
```

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+* This function applies damage to internal organs in case the mob died having toxin damage
+* This feature increases amount of effort it takes to revive a victim of poisoning
+*/
+/mob/living/carbon/proc/process_toxin_damage_on_death(mob/living/carbon/target)
```
```suggestion
/mob/living/carbon/proc/process_toxin_damage_on_death(mob/living/carbon/target)
if (!target)
	return
	
```

Gaxeer commented on code/modules/mob/living/carbon/death.dm:
```diff
@@ -1,3 +1,44 @@
+		if (organ_picked.damage >= organ_picked.max_damage) // you can't check if the organ is dead because of immortality of the brain
+			to_be_picked_from.Remove(organ_picked)
+
+
```
```suggestion
```

fake123acc committed:
Update code/modules/mob/living/carbon/death.dm

Idk i guess its somewhat better

Co-authored-by: Gaxeer <44334376+Gaxeer@users.noreply.github.com>

fake123acc committed:
Update code/modules/mob/living/carbon/death.dm

Added if(!target) check

Co-authored-by: Gaxeer <44334376+Gaxeer@users.noreply.github.com>

fake123acc committed:
Update code/modules/mob/living/carbon/death.dm

Cleaned extra \n

Co-authored-by: Gaxeer <44334376+Gaxeer@users.noreply.github.com>

Bizzonium commented:
Всё же надо обозначить границы для while

fake123acc commented:
> Всё же надо обозначить границы для while

Это же хак, чтобы макросом можно было заменить функцию, разве нет?
Или речь про другой while, где мы урон по органам раздаем? Ну, там вроде границы определены. Поставить счётчик-костыль, чтобы после Х попыток принудительно выбрасывал из цикла?
